File: complexipy/utils/ignored.py
from __future__ import annotations

import json
import os
from typing import List, Optional

from rich.console import Console

from complexipy import _complexipy
from complexipy.types import OutputFormat
from complexipy.utils.paths import resolve_output_paths


def handle_report_ignored(
    console: Console,
    report_ignored: bool,
    paths: Optional[List[str]],
    exclude: Optional[List[str]],
    output_formats: List[OutputFormat],
    output: Optional[str],
    no_ignore: bool,
    invocation_path: str,
) -> None:
    if not report_ignored:
        return

    paths = paths or []
    exclude = exclude or []
    ignored_locations, _ = _complexipy.collect_all_ignored_locations(
        paths, exclude, invocation_path
    )
    if ignored_locations:
        for loc in ignored_locations:
            console.print(f"{loc.path}:{loc.line}  {loc.comment}")
        console.print(
            f"\nFound {len(ignored_locations)} suppressed location(s)."
        )
        if no_ignore:
            console.print("(all markers ignored due to --no-ignore)")
    else:
        console.print("No ignore comments found.")

    if OutputFormat.json in output_formats and ignored_locations:
        ignored_output_paths = resolve_output_paths(
            [OutputFormat.json], output, invocation_path
        )
        ignored_json_path = os.path.join(
            os.path.dirname(ignored_output_paths[OutputFormat.json]),
            "complexipy-ignored.json",
        )
        ignored_data = [
            {"path": loc.path, "line": loc.line, "comment": loc.comment}
            for loc in ignored_locations
        ]
        with open(ignored_json_path, "w") as f:
            json.dump(ignored_data, f, indent=2)
            f.write("\n")
        console.print(f"Ignored locations saved at {ignored_json_path}")

File: crates/complexipy-cli/src/utils/ignored.rs
use std::fs;

use crate::types::OutputFormat;
use crate::utils::paths::resolve_output_paths;
use complexipy_core::classes::{IgnoredLocation, RemovableIgnore};
use complexipy_core::runner::{
    collect_all_ignored_locations_shared, collect_removable_ignored_locations_shared,
};
use complexipy_core::utils::ExportError;

pub fn handle_report_ignored(
    report_ignored: bool,
    paths: &[String],
    exclude: &[String],
    output_formats: &[OutputFormat],
    output: Option<&str>,
    _no_ignore: bool,
    invocation_path: &str,
) -> Result<(Vec<IgnoredLocation>, Option<String>), ExportError> {
    if !report_ignored {
        return Ok((Vec::new(), None));
    }

    let (ignored_locations, _) =
        collect_all_ignored_locations_shared(paths, exclude, invocation_path)
            .map_err(ExportError::Io)?;

    let ignored_json_path =
        if output_formats.contains(&OutputFormat::Json) && !ignored_locations.is_empty() {
            let ignored_output_paths = resolve_output_paths(
                &[OutputFormat::Json],
                output,
                std::path::Path::new(invocation_path),
            )
            .map_err(|e| ExportError::Io(e.to_string()))?;
            let ignored_output_path = &ignored_output_paths[0].1;
            let ignored_dir = std::path::Path::new(ignored_output_path)
                .parent()
                .unwrap_or(std::path::Path::new("."));
            let ignored_json_path = ignored_dir.join("complexipy-ignored.json");

            let ignored_data: Vec<serde_json::Value> = ignored_locations
                .iter()
                .map(|location| {
                    serde_json::json!({
                        "path": location.path,
                        "line": location.line,
                        "comment": location.comment,
                    })
                })
                .collect();
            let serialized = serde_json::to_string_pretty(&ignored_data).map_err(|e| {
                ExportError::Serialize(format!("Failed to serialize ignored locations: {}", e))
            })?;
            fs::write(&ignored_json_path, format!("{}\n", serialized)).map_err(|e| {
                ExportError::Io(format!(
                    "Failed to write ignored locations to {}: {}",
                    ignored_json_path.display(),
                    e
                ))
            })?;

            Some(ignored_json_path.to_string_lossy().into_owned())
        } else {
            None
        };

    Ok((ignored_locations, ignored_json_path))
}

pub fn handle_removable_ignores(
    paths: &[String],
    exclude: &[String],
    max_complexity_allowed: u64,
    invocation_path: &str,
) -> Vec<RemovableIgnore> {
    collect_removable_ignored_locations_shared(
        paths,
        exclude,
        max_complexity_allowed,
        invocation_path,
    )
    .map(|(removable_ignores, _)| removable_ignores)
    .unwrap_or_default()
}

#[cfg(test)]
mod tests;
